recipe/meta.yaml

```diff
@@ -49,18 +45,84 @@ requirements:
     - pytorch
     - pytorch =*=cuda*
     - setuptools
-  run:
-    - einops
-    - python
-    - pytorch =*=cuda*
 
-test:
-  imports:
-    - flash_attn
-  commands:
-    - pip check
-  requires:
-    - pip
+outputs:
+  - name: flash-attn
+    requirements:
+      build:
+        - {{ compiler('cxx') }} # needed for DSO checker
+        - {{ stdlib('c') }}     # needed for DSO checker
+      host:
+        - cuda-version {{ cuda_compiler_version }}  # same cuda for host and build
+        - cuda-cudart-dev  # [(cuda_compiler_version or "").startswith("12")]
+        - python
+        - pytorch =*=cuda*
+      run:
+        - einops
+
+    files:
+      include:
+        - 'lib/python*/site-packages/flash_attn/**'
+        - 'lib/python*/site-packages/flash_attn-{{ version }}.dist-info/**'
+        - 'lib/python*/site-packages/flash_attn_2_cuda.cpython-*.so'
+
+    test:
+      imports:
+        - flash_attn
+      commands:
+        - pip check
+      requires:
+        - pip
+
+  - name: flash-attn-fused-dense-lib
+    requirements:
+      build:
+        - {{ compiler('cxx') }} # needed for DSO checker
+        - {{ stdlib('c') }}     # needed for DSO checker
+      host:
+        - cuda-version {{ cuda_compiler_version }}  # same cuda for host and build
+        - cuda-cudart-dev  # [(cuda_compiler_version or "").startswith("12")]
+        - python
+        - pytorch =*=cuda*
+      run:
+        - {{ pin_subpackage('flash-attn', exact=True) }}
+
+    files:
+      include:
+        - 'lib/python*/site-packages/fused_dense_lib.cpython-*.so'
+
+    test:
+      imports:
+        - flash_attn.ops.fused_dense
+      commands:
+        - pip check
+      requires:
+        - pip
+
+  - name: flash-attn-layer-norm
```

Comment:
Are these package names upstream canonical? Otherwise, why does one match the extension module name and the other does not? Can we use the import names? i.e. `layer-norm` and `fused-dense`?

Reply:
There are two sets of names. The extension directories:
* `csrc/fused_dense_lib`
* `csrc/layer_norm`

Then the python packages:
* `fused_dense_lib`
* `dropout_layer_norm`

The two imports are the users of these extensions:
* In `flash_attn.ops.fused_dense`: `import fused_dense_lib as fused_dense_cuda`
* In `flash_attn.ops.layer_norm`: `import dropout_layer_norm`

I'm not sure which set of names we should use.

Reply:
I don't see "fused_dense_lib" on PYPI. https://pypi.org/search/?q=fused_dense_lib&o=

Reply:
Dropout layer norm is deprecated. https://github.com/Dao-AILab/flash-attention/blob/main/csrc/layer_norm/README.md